tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestTimeCycle.java

```diff
@@ -0,0 +1,88 @@
+package org.unicode.cldr.unittest;
+
+import com.ibm.icu.text.DateTimePatternGenerator;
+import com.ibm.icu.text.SimpleDateFormat;
+import com.ibm.icu.util.ULocale;
+import java.util.logging.Logger;
+import org.junit.jupiter.api.Test;
+import org.unicode.cldr.icu.LDMLConstants;
+import org.unicode.cldr.util.CLDRConfig;
+import org.unicode.cldr.util.CLDRLocale;
+import org.unicode.cldr.util.Factory;
+import org.unicode.cldr.util.ICUServiceBuilder;
+
+public class TestTimeCycle {
+    private static final Logger logger = Logger.getLogger(TestTimeCycle.class.getName());
+
+    /**
+     * Test that a locale's standard time format matches the region's preferred time cycle
+     *
+     * <p>This is similar to icu4j DateTimeGeneratorTest.testJjMapping. It uses current CLDR data,
+     * not ICU data.
+     */
+    @Test
+    public void testJjMapping() {
+        final String jSkeleton = "j";
+        final char[] timeCycleChars = {'H', 'h', 'K', 'k'};
+        // First test that j maps correctly by region in a locale for which we do not have data.
+        {
+            String testLocaleID = "de_US"; // short patterns from fallback locale "de" have "HH"
+            ULocale testLocale = new ULocale(testLocaleID);
+            DateTimePatternGenerator dtpg = DateTimePatternGenerator.getInstance(testLocale);
+            String jPattern = dtpg.getBestPattern(jSkeleton);
+            String jPatSkeleton = dtpg.getSkeleton(jPattern);
+            if (jPatSkeleton.indexOf('h')
+                    < 0) { // expect US preferred cycle 'h', not H or other cycle
+                logger.severe(
+                        "DateTimePatternGenerator getBestPattern locale "
+                                + testLocaleID
+                                + ", pattern j did not use 'h'");
+            }
+        }
+
+        // Next test that in all available Locales, the actual short time pattern uses the same
+        // cycle as produced by 'j'
+
+        CLDRConfig testInfo = CLDRConfig.getInstance();
+        Factory cldrFactory = testInfo.getCldrFactory();
+        for (String localeID : cldrFactory.getAvailable()) {
+            CLDRLocale loc = CLDRLocale.getInstance(localeID);
+            ICUServiceBuilder icuServiceBuilder = ICUServiceBuilder.forLocale(loc);
+
+            // Compare ICU data version:
+            // DateTimePatternGenerator dtpg = DateTimePatternGenerator.getInstance(uloc);
+
+            DateTimePatternGenerator dtpg = DateTimePatternGenerator.getEmptyInstance();
```

Comment:
`getEmptyInstance`? Is it OK that `dtpg` itself isn't associated with the locale? Instead it's `icuServiceBuilder` that's locale-specific.

Reply:
You will need to load up the DateTimePatternGenerator yourself with the CLDR data for the locale, it does not automatically do that. Take a look at tools/cldr-code/src/main/java/org/unicode/cldr/test/FlexibleDateFromCLDR.java (in fact you may be able to use that class in this test)